crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -202,155 +437,43 @@ trait Inner: StateWrite {
-            step_size,
-            next_step_index
-        )
-    })?;
-
-    Ok(start_height.checked_add(next_step_size_from_start))
-}
-struct TriggerData {
-    pub start_height: u64,
-    pub end_height: u64,
-    pub current_height: u64,
-    pub step_count: u64,
-}
-
-#[cfg(test)]
-mod tests {
-    use crate::component::dutch_auction::{compute_next_trigger, TriggerData};
-
-    #[test]
-    fn test_current_height_equals_start_height() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 100,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), Some(120));
-    }
-
-    #[test]
-    fn test_current_height_equals_end_height() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 200,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), None);
-    }
-
-    #[test]
-    fn test_current_height_below_start_height() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 90,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), Some(120));
-    }
-
-    #[test]
-    fn test_current_height_above_end_height() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 210,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), None);
-    }
-
-    #[test]
-    fn test_current_height_below_boundary() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 119,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), Some(120));
-    }
-
-    #[test]
-    fn test_current_height_above_boundary() {
-        let data = TriggerData {
-            start_height: 100,
-            end_height: 200,
-            step_count: 5,
-            current_height: 121,
-        };
-        assert_eq!(compute_next_trigger(data).unwrap(), Some(140));
-    }
+/// TODO(erwan): prove that this works. More work needed to work out validation
+/// rules that let us operate within the 112 bits constraint.
+fn compute_pq_at_step(
+    auction_description: &DutchAuctionDescription,
+    step_index: u64,
+) -> (Amount, Amount) {
+    let step_index = Amount::from(step_index);
+    let step_count = Amount::from(auction_description.step_count);
+
+    let q = auction_description
+        .input
+        .amount
+        .checked_mul(&step_count)
+        // TODO(erwan): this isn't true, this must part of 112 bits extra validation pass todo.
+        .expect("checked in validation");
+
+    let p = (step_count - step_index) * auction_description.max_output
+        + step_index * auction_description.min_output;
```

Comment:
I think this could be a little clearer if it were factored like
```
// The target output, scaled up by `step_count` to avoid divisions.
// Linearly interpolate between `max_output` at `step_index = 0`
//                          and `min_output` at `step_index = step_count`.
let target_output_scaled = (step_count - step_index) * max_output + step_index * min_output;
// The input, scaled up by `step_count` to match.
let input_scaled = step_count * input.amount;

// The trading function interpolates between (input, 0) and (0, target_output)
let p = target_output_scaled;
let q = target_input_scaled;
```

Reply:
Also doesn't this mean we need to iterate up to and INCLUDING step_count?

Reply:
Thanks, sg

Reply:
I just thought about it, and I think we need to change the interpolation so that:
- the minimum price is hit at the penultimate trigger: `step_count - 1`
- the auction is closed at the ultimate trigger: `step_count` (= `end_height`)

This means an auction has at least 2 steps, that seems like a reasonable tradeoff and already baked in with the idea of a `min` and `max` outputs (two points).